google_guest_agent/addresses_windows.go

```diff
@@ -120,22 +121,120 @@ type (
 )
 
 func addAddress(ip net.IP, mask net.IPMask, index uint32) error {
+	logger.Infof("Adding %q ip address", ip.String())
+	subnet, _ := mask.Size()
 	// CreateUnicastIpAddressEntry only available Vista onwards.
-	if err := procCreateUnicastIpAddressEntry.Find(); err != nil {
+	// AddIPAddress api supports only ipv4 address.
+	if isIPv6(ip) {
+		return createUnicastIpAddressEntry6(ip, uint8(subnet), index)
+	}
+
+	// https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-addipaddress
+	if err := procCreateUnicastIpAddressEntry.Find(); err != nil && !isIPv6(ip) {
 		return addIPAddress(ip, mask, index)
 	}
-	subnet, _ := mask.Size()
 	return createUnicastIpAddressEntry(ip, uint8(subnet), index)
 }
 
-func removeAddress(ip net.IP, index uint32) error {
+func removeAddress(ip net.IP, mask net.IPMask, index uint32) error {
+	logger.Infof("Removing %q ip address", ip.String())
+	subnet, _ := mask.Size()
+	if isIPv6(ip) {
+		// Unlike ipv4 that can be added either by addIPAddress or createUnicastIpAddressEntry
+		// ipv6 addresses can only be added by createUnicastIpAddressEntry6.
+		// Try removing them by deleteUnicastIpAddressEntry6 only as deleteIPAddress deletes
+		// IP address previously added using AddIPAddress only.
+		return deleteUnicastIpAddressEntry6(ip, uint8(subnet), index)
+	}
+
 	// DeleteUnicastIpAddressEntry only available Vista onwards.
 	if err := procDeleteUnicastIpAddressEntry.Find(); err != nil {
 		return deleteIPAddress(ip)
 	}
 	return deleteUnicastIpAddressEntry(ip, index)
 }
 
+// LUID represents the locally unique identifier (LUID) for a network interface.
+// https://learn.microsoft.com/en-us/windows/win32/api/ifdef/ns-ifdef-net_luid_lh
+type LUID uint64
+
+// RawSockaddrInet represents an IPv4/IPv6 address and family.
+// https://learn.microsoft.com/en-us/windows/win32/api/ws2ipdef/ns-ws2ipdef-sockaddr_inet
+type RawSockaddrInet struct {
+	Family uint16
+	data   [26]byte
+}
+
+// MIB_UNICASTIPADDRESS_ROW6 stores unicast IP address information used with for
+// ipforwarding addresses.
+// https://learn.microsoft.com/en-us/windows/win32/api/netioapi/ns-netioapi-mib_unicastipaddress_row
+type MIB_UNICASTIPADDRESS_ROW6 struct {
```

Comment:
No underscores in go names, `MIBUnicastIPAddressRow6`

Reply:
done